hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableECContainerProvider.java

```diff
@@ -92,12 +92,16 @@ public WritableECContainerProvider(ConfigurationSource conf,
   public ContainerInfo getContainer(final long size,
       ECReplicationConfig repConfig, String owner, ExcludeList excludeList)
       throws IOException {
+    // Bound this at a minimum of 1 byte in case a request is made for a very
+    // small size, which when divided by EC DataNum is zero.
+    long requiredSpace = Math.max(1, size / repConfig.getData());
```

Comment:
Hi Stephen, I'm just to confirm that whether it is intended to get the floor of (size / repConfig.getData()) ?

Reply:
`size` and `repConfig.getData()` are long and int, so `1 / 6`, for example, will give zero. In that case, I wanted to check there is 1 byte of space rather than zero bytes, which is why I had `Math.max(1, size / data)` - the idea is to prevent us getting a zero here.

Reply:
The value for the space used in the container here is approximate and always lagging as it does not have the current view of the usage from the DNs. Integer division always gives the floor I think, so we are always rounding down. What do you think the calculation should be?

Reply:
Hmmm, the basic idea is good to me, but there are things to consider on the edge.
- If we use a floor here, then we'll get less chance to close the container actively right after the check - `containerHasSpace` and at the same time slightly greater chance to get a passive full event from DN, because the block allocation succeeded, and client continues to write more bytes to the container and will potentially close the container as it is full. And then client retries write and a new container may be allocated if needed.
- If we use a slightly larger value > (size/data)(e.g. the ceil), then we'll get greater chance to actively close the container at SCM side, and a new container may be allocated right now for the client so no retry needed, but the container usedBytes will be less close to 5GB.

So, IMHO, is a value like the ceil more suitable as it may save a client retry?
Please correct me if I was wrong.

Reply:
I don't think it really matters, but I can change it to ceil if you like.

The DN container size limit is not a hard value. When the limit is triggered, the DN signals to SCM to close the containers and SCM closes the container, but the DNs give a grace period to allow inflight writes to continue. That should avoid clients needing to retry.

The current container usage in SCM is not accurate - it is an estimate and updated periodically.

We also don't know how large the block is going to be - it might only actually write 1MB of data - there is no way to know.

I have considered whether the logic here should trigger the container to close or not. I was mostly convinced it does not belong here, and we should let the DNs trigger the close, but I worried about us ending up with many pipelines we have to iterate over to find one with space, slowing down the block allocation.

Reply:
LGTM+1